includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign-usage-reports.php

```diff
@@ -0,0 +1,193 @@
+		$cutoff_datetime = strtotime( '-' . $last_n_days . ' days' );
+		foreach ( $all_contacts as $contact ) {
+			if ( isset( $contact[ $param ] ) ) {
+				$date        = strtotime( $contact[ $param ] );
+				$date_string = gmdate( 'Y-m-d', $date );
+				if ( $date < $cutoff_datetime ) {
+					continue;
+				}
+				if ( isset( $report[ $date_string ] ) ) {
+					if ( isset( $report[ $date_string ][ $report_key ] ) ) {
+						$report[ $date_string ][ $report_key ]++;
+					} else {
+						$report[ $date_string ][ $report_key ] = 1;
+					}
+				} else {
+					$report[ $date_string ] = [
+						$report_key => 1,
+					];
+				}
+			}
+		}
+		return $report;
+	}
+
+	/**
+	 * Get contacts data - subs and unsubs.
+	 *
+	 * @param int $last_n_days Number of last days to get the data about.
+	 */
+	private static function get_contacts_data( $last_n_days ) {
+		$report = [];
+
+		$cutoff_datetime = strtotime( '-' . $last_n_days . ' days' );
+
+		// Subscribers, with the cutoff date – only created (subscribed) afterwards.
+		$subscribed = self::get_all_contacts( 1, $cutoff_datetime );
+		if ( \is_wp_error( $subscribed ) ) {
+			return $subscribed;
+		}
+		$report = self::update_report_with_contact_data( 'cdate', 'subs', $subscribed, $report, $last_n_days );
+
+		// Unsubscribed contacts, without the cutoff date. All have to be pulled because
+		// there's no way to filter by unsubscribed date.
+		$unsubscribed_contacts = self::get_all_contacts( 2 );
+		if ( \is_wp_error( $unsubscribed_contacts ) ) {
+			return $unsubscribed_contacts;
+		}
+		$report = self::update_report_with_contact_data( 'udate', 'unsubs', $unsubscribed_contacts, $report, $last_n_days );
+
+		return $report;
+	}
+
+	/**
+	 * Get campaign data - emails sent, opens, and clicks.
+	 *
+	 * @param int $last_n_days Number of last days to get the data about.
+	 */
+	private static function get_campaign_data( $last_n_days ) {
+		$report = [];
+
+		$ac               = self::get_ac_instance();
+		$params           = [
+			'query' => [
+				'limit'  => 100, // Assuming there will be no more than 100 campaigns in the requested period (last n days).
+				'orders' => [ 'sdate' => 'DESC' ],
+			],
+		];
+		$campaigns_result = $ac->api_v3_request( 'campaigns', 'GET', $params );
+		$cutoff_datetime  = strtotime( '-' . $last_n_days . ' days' );
+
+		if ( \is_wp_error( $campaigns_result ) ) {
+			return $campaigns_result;
+		}
+		foreach ( $campaigns_result['campaigns'] as $campaign ) {
+			if ( ! isset( $campaign['sdate'] ) ) {
+				continue;
+			}
+			// If the send date is before the cutoff date, skip.
+			$campaign_send_date = strtotime( $campaign['sdate'] );
+			if ( $campaign_send_date < $cutoff_datetime ) {
+				continue;
+			}
+			$report[ gmdate( 'Y-m-d', $campaign_send_date ) ] = [
+				'emails_sent' => $campaign['send_amt'],
+				'opens'       => $campaign['uniqueopens'],
+				'clicks'      => $campaign['uniquelinkclicks'],
```

Comment:
I think we are not looking at the right thing here.

You are looking at the day that the campaign was sent, and then getting the number of opens and clicks.

I think what we expect is the number of opens and clicks across all campaigns in a given day. (and it could be an open of a campaign that was sent before the cutoff date).

Let's discuss it

Reply:
You're correct, if more than one campaign was sent on a day, the stats would be overriden with the latest campaign. I'll update it.

Reply:
Not only that. We are not looking at the opens at that day, we are looking at opens of campaigns sent that day, which are different things

Reply:
Documenting what we discussed:

- Here we want to store the number of emails sent on that day.

On another check, we will loop through campaigns sent in the last 30 days and check for the number of opens and clicks
- https://developers.activecampaign.com/reference/retrieve-a-campaign
- We will try to, as close as midnight as possible, get the totals and store them - then do the same thing the next day to get the number of opens and clicks for that particular day

Reply:
ah, and we can also drop the support to get data from more than one day in the past

Reply:
> We are not looking at the opens at that day, we are looking at opens of campaigns sent that day, which are different things

I can't see an API endpoint to retrieve per-day data about opens and clicks, so we'll have to assume that the opens & clicks happen on the date when the campaign is sent. There are such endpoints in the V1 API ([campaign_report_open_list](https://www.activecampaign.com/api/example.php?call=campaign_report_open_list), [campaign_report_link_list](https://www.activecampaign.com/api/example.php?call=campaign_report_link_list)), but I've found them unreliable, often returning no data. 
IMO the exact dates of the opens & clicks are not very relevant – conversion rate per campaign, based on per-campaign numbers, is the important metric to pay attention to when evaluating campaign success.

Reply:
But if we do that we'll have to keep checking for the past campaigns and update the data fro the previous days that are already in BQ. 

We'll need to check the campaigns details totals, store the value and subtract the number we had yesterday (https://href.li/?https://developers.activecampaign.com/reference/list-all-campaigns)

Reply:
Hm, same goes for Mailchimp implementation, doesn't it? The activity response we're reading is for yesterday, and a campaign might still be opened tomorrow.

Reply:
No, we are precisely trying to get the same result we have with Mailchimp.

Mailchimp gives us the total number of opens/clicks for each day across all campaigns.

> The activity response we're reading is for yesterday, and a campaign might still be opened tomorrow.

No problem. Tomorrow we'll get them.

Reply:
You're right. With Mailchimp there will be no missed data, while with the current approach to AC here, if a campaign is opened a day after data was fetched, this fact won't make it to the report. 

Taking into account that we can only get totals from the API, a way to get per-day values would be:
1. Fetch data and store it locally (e.g. campaign A had 10 total opens last time)
2. Compare with stored data and subtract it from new data (e.g. campaign A has 15 total opens now, so it must've gotten opened 5 times since the last report)

This way we would not have to update historical data in BQ. I've added some tests so it's all clearer.